### ISSUE TITLE
Having flag attributes is cumbersome with current API

### ISSUE THREAD

**Issue by panglesd** [ocaml-ppx/ppxlib#404]
With current API, one needs to create an "empty payload" pattern:
```ocaml
let p = Ast_pattern.(pstr nil)
```
then register the attribute with a `unit` continuation:
```ocaml
let attr = Attribute.declare name context p ()
```
And finally, when getting it, turn the `unit option` type into a `bool` type:
```ocaml
match Attribute.get attr x with
    | Some () -> true
    | None -> false
```

All of this could be simplified with the following API addition:
```ocaml
val declare_flag : string -> 'a Context.t -> ('a, unit) t
val has_flag : ('a, unit) t -> ?mark_as_seen:bool -> 'a -> bool
```

(reported by @sim642 in https://github.com/ocaml-ppx/ppx_deriving/pull/263#discussion_r924295740)

**Comment by dianaoigo** [comment]
@panglesd i would like to work on this issue

**Comment by dianaoigo** [comment]
@panglesd i have forked and cloned the project . looking at the project, would it be a good idea to make the suggested changes in the src/ast_pattern.ml file? I am still learning and understanding more about the ocaml ecosystem.

**Comment by panglesd** [comment]
Attributes are part of the [OCaml syntax](https://v2.ocaml.org/releases/5.0/manual/attributes.html), as in:
```ocaml
type t = {
  a : int [@foo "an attribute named foo whose payload is a string"]
}
```

`Ast_pattern` is a module to build "pattern-matches of the AST". It is used in several places in ppxlib, mostly to extract payloads in extension nodes and attributes. The documentation on Ast_pattern is available [here](https://ocaml.org/p/ppxlib/latest/doc/matching-code.html#ast_pattern_intro).

`ppxlib` allows "specifying" attributes and their payloads, from a name, a context and an ast pattern. This is the [`declare`](https://ocaml.org/p/ppxlib/latest/doc/Ppxlib/Attribute/index.html#val-declare) function. From a declared attribute, one can then extract the value in the payload, using the ast-pattern: this is [`get`](https://ocaml.org/p/ppxlib/latest/doc/Ppxlib/Attribute/index.html#val-get).

Here, the issue is about adding support for a common case, where the attributes have no payload, the only thing we are interested in is whether they are present or not. For this, I propose to add the `declare_flag` and `has_flag` function. According to you, in which module in the API should they belong?

**Comment by dianaoigo** [comment]
@panglesd  i don't know why i haven't been able to see till i restarted my machine today. At some point i thought you were ignoring me. I was actually caught between the ast_pattern and attrribute model i just wasn't sure which. for now i believe attribute module would be the best way since it's where declaration and retrieval of attributes is done.

**Comment by dianaoigo** [comment]
@panglesd since i restarted my machine i am having serious problems. now when i run _dune build _ i get the error 
File "dune-project", line 6, characters 1-5:
6 | (cram enable)
     ^^^^
Error: Unknown field cram 
and when i try reinstalling the dependencies using _opam install --deps-only --with-test .
_ i get the error 
[ERROR] Package conflict!
  * Missing dependency:
    - ppxlib < 0.1.0
    no matching version

No solution found, exiting
 i am using ubuntu

**Comment by dianaoigo** [comment]
@pitag-ha i am having some problems. i restarted my machine after almost a week and it must have corrupted everything. now i can even build or test as shown in the comment above

**Comment by panglesd** [comment]
Hello @dianaoigo !
First, if you ever feel that I am ignoring you, please ping me. It could happen that I miss one gh notification.
This was not the case here: I thought you were busy working on this or some other project since you did not reply to my clarifications.

The error message when running `dune build` seems to indicate that `dune` has a too low version. Reinstalling the dependencies was a good move, but the error you get is strange (I think I already had it once, I think it is due to `ppxlib-bench.opam`).

You could try:
- Rerunning your install command, after having removed `ppxlib-bench.opam` (or only this line: https://github.com/ocaml-ppx/ppxlib/blob/main/ppxlib-bench.opam#L17)
- Install dependencies by hand, see the list of them https://github.com/ocaml-ppx/ppxlib/blob/main/ppxlib.opam#L22..L35

Let me know if that unblocks the situation!

**Comment by panglesd** [comment]
Also, it is strange that this happened after a machine reboot... Maybe you changed the opam switch, or your environment is not up to date with the switch you have chosen? (``eval $(opam env)``)

**Comment by dianaoigo** [comment]
I uninstalled everything dune, opam ocaml, reinstalled and updated. somehow that solved the issue. also learnt that any  dune version below 2.7 will throw some crazy errors. I have been working on a different ocaml project maybe that was the cause. Could you offer a little more insight on how to proceed?

**Comment by panglesd** [comment]
@dianaoigo I already gave all the generic information I could in https://github.com/ocaml-ppx/ppxlib/issues/404#issue-1637106035 and https://github.com/ocaml-ppx/ppxlib/issues/404#issuecomment-1482812845, in order to help you more I would need a more detailed question, for instance I need to know what is clear and what is unclear yet. There is a nice blog post about asking good questions [here](https://jvns.ca/blog/good-questions/) that explains that very well!

**Comment by sim642** [comment]
This should be closed by #408 I think.
